airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java

```diff
@@ -104,101 +118,249 @@ public AirbyteConnectionStatus check(JsonNode config) {
     }
   }
 
-  protected String getCurrentDatabaseName(DSLContext ctx) {
-    return ctx.select(currentSchema()).fetch().get(0).getValue(0, String.class);
-  }
-
   @Override
   public AirbyteCatalog discover(JsonNode config) throws Exception {
-    try (final Database database = createDatabase(config)) {
+    try (final JdbcDatabase database = createDatabase(config)) {
       return new AirbyteCatalog()
-          .withStreams(getTables(database)
-              .stream()
-              .map(t -> CatalogHelpers.createAirbyteStream(t.getName(), t.getFields()))
-              .collect(Collectors.toList()));
+          .withStreams(getTables(
+              database,
+              Optional.ofNullable(config.get("database")).map(JsonNode::asText),
+              Optional.ofNullable(config.get("schema")).map(JsonNode::asText))
+                  .stream()
+                  .map(t -> CatalogHelpers.createAirbyteStream(t.getName(), t.getFields())
+                      .withSupportedSyncModes(Lists.newArrayList(SyncMode.FULL_REFRESH, SyncMode.INCREMENTAL)))
+                  .collect(Collectors.toList()));
     }
   }
 
-  protected List<String> getExcludedInternalSchemas() {
-    return List.of();
-  }
+  @Override
+  public Stream<AirbyteMessage> read(JsonNode config, ConfiguredAirbyteCatalog catalog, JsonNode state) throws Exception {
+    final JdbcStateManager stateManager = new JdbcStateManager(state == null ? new JdbcState() : Jsons.object(state, JdbcState.class), catalog);
```

Comment:
NIT: for the empty `JdbcState`, I recommend that you create a static function in the class that document what your doing:
`JdbcState.noState()` or `JdbcState.empty()`

That will also make it easier in the future to change what an empty state is without having to change all the calling sites.

Reply:
unfortunately `JdbcState` is generated code. I can put that static on the manager though. i'll do that.